packages/client/src/pages/Game/engine/TDEngine.tsx

```diff
@@ -0,0 +1,770 @@
+      this.isCanBuild = true;
+      this.draftTower = new Tower(
+        this,
+        this.towerSprites.levelTwo!,
+        this.projectileSprites.levelTwo!,
+        this.projectileHitSprites.levelTwo!,
+        this.draftBuildCoordinates,
+        this.towerTwoParam.towerParams,
+        this.towerTwoParam.projectileParams,
+      );
+    }
+  };
+
+  public buildThirdTower = () => {
+    if (this.isEnoughMoney(this.towerThreeParam.towerParams.price)) {
+      this.isCanBuild = true;
+      this.draftTower = new Tower(
+        this,
+        this.towerSprites.levelThree!,
+        this.projectileSprites.levelThree!,
+        this.projectileHitSprites.levelThree!,
+        this.draftBuildCoordinates,
+        this.towerThreeParam.towerParams,
+        this.towerThreeParam.projectileParams,
+      );
+    }
+  };
+
+  public findClosestTile(coordinates: ITwoDCoordinates) {
+    let minDistance = this.map?.mapParams.width;
+    for (let tile of this.map?.mapParams.mapTilesArr!) {
+      const distance =
+        (tile.x -
+          coordinates.x! +
+          this.map?.mapParams.gridStep! -
+          this.map?.mapParams?.tileCenter!) *
+          (tile.x -
+            coordinates.x! +
+            this.map?.mapParams.gridStep! -
+            this.map?.mapParams?.tileCenter!) +
+        (tile.y -
+          coordinates.y! +
+          this.map?.mapParams.gridStep! +
+          this.map?.mapParams?.tileCenter!) *
+          (tile.y -
+            coordinates.y! +
+            this.map?.mapParams.gridStep! +
+            this.map?.mapParams?.tileCenter!);
+      if (distance < minDistance!) {
+        minDistance = distance;
+        this.map!.mapParams.closestTile! = tile!;
+      }
+    }
+
+    this.draftBuildCoordinates = {
+      x: this.map?.mapParams.closestTile.x! + this.map?.mapParams.gridStep!,
+      y: this.map?.mapParams.closestTile.y! + this.map?.mapParams.gridStep!,
+    };
+
+    return {
+      x: this.map?.mapParams.closestTile.x,
+      y: this.map?.mapParams.closestTile.y!,
+    };
+  }
+
+  highlightTile(coords: ITwoDCoordinates) {
+    this.context?.beginPath();
+    this.context!.strokeStyle = "green";
+    this.context?.setLineDash([]);
+    this.context?.strokeRect(
+      coords.x,
+      coords.y,
+      this.map?.mapParams.gridStep!,
+      this.map?.mapParams.gridStep!,
+    );
+    this.context?.closePath();
+  }
+
+  public canvasMouseMoveCallback = (e: MouseEvent) => {
+    this.cursorPosition = { x: e.pageX, y: e.pageY };
+    this.findClosestTile(this.cursorPosition);
+    this.draftShowTower();
+  };
+
+  public canvasClickCallback = (e: MouseEvent) => {
+    this.draftBuildTower();
+  };
+
+  public gameWindowKeydown = (e: KeyboardEvent) => {
+    this.manageHotkeys(e);
+  };
+
+  public isEnoughMoney(towerPrice: ITower["towerParams"]["price"]) {
+    if (this.money >= towerPrice!) {
+      return true;
+    }
+    this.isNotEnoughMoney = true;
```

Comment:
and if there is enough money, why you don't switch `this.isNotEnoughMoney` back?)
```suggestion
    if (this.money >= towerPrice!) {
      this.isNotEnoughMoney = false;
      return true;
    }
    this.isNotEnoughMoney = true;
```

Reply:
Skipped, tower price is not constant and may vary for other towers, so if you can afford first tower you may have not enough money to build third tower